# 01. Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Project folder pathway

path = r'C:\Users\SD.DESKTOP-45BHU58\OneDrive\Desktop\CF\Section 6 - Advanced Analytics\Final Project'

In [3]:
path

'C:\\Users\\SD.DESKTOP-45BHU58\\OneDrive\\Desktop\\CF\\Section 6 - Advanced Analytics\\Final Project'

In [4]:
# Importing housing data

df_house = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'austinHousingData.csv'))

# 03. Data Wrangling

In [6]:
# Checking data shape

df_house.shape

(15171, 47)

## a. Data Types

In [7]:
# Checking data types

df_house.dtypes

zpid                            int64
city                           object
streetAddress                  object
zipcode                         int64
description                    object
latitude                      float64
longitude                     float64
propertyTaxRate               float64
garageSpaces                    int64
hasAssociation                   bool
hasCooling                       bool
hasGarage                        bool
hasHeating                       bool
hasSpa                           bool
hasView                          bool
homeType                       object
parkingSpaces                   int64
yearBuilt                       int64
latestPrice                     int64
numPriceChanges                 int64
latest_saledate                object
latest_salemonth                int64
latest_saleyear                 int64
latestPriceSource              object
numOfPhotos                     int64
numOfAccessibilityFeatures      int64
numOfApplian

In [8]:
# Checking values for lotSizeSqFt

df_house['lotSizeSqFt'].value_counts(dropna = False)

11325.6    300
12196.8    285
11761.2    281
6969.0     258
7840.0     233
          ... 
9718.0       1
55756.8      1
10863.0      1
94089.6      1
3097.0       1
Name: lotSizeSqFt, Length: 1426, dtype: int64

In [9]:
# Checking values for livingAreaSqFt

df_house['livingAreaSqFt'].value_counts(dropna = False)

2068    30
1200    29
1590    28
2035    28
1100    27
        ..
4455     1
6706     1
7600     1
3197     1
6068     1
Name: livingAreaSqFt, Length: 3650, dtype: int64

In [10]:
# Changing data type of lotSizeSqFt from float to int to keep square feet format consistent

df_house['lotSizeSqFt'] = df_house['lotSizeSqFt'].astype('int64')

In [11]:
# Re-checking values for lotSizeSqFt

df_house['lotSizeSqFt'].value_counts(dropna = False)

11325    300
12196    285
11761    281
6969     258
7840     233
        ... 
9718       1
55756      1
10863      1
94089      1
3097       1
Name: lotSizeSqFt, Length: 1426, dtype: int64

In [12]:
# Checking all columns

pd.options.display.max_columns = None
df_house.head()

zpid          city          streetAddress  zipcode  \
0   111373431  pflugerville   14424 Lake Victor Dr    78660   
1   120900430  pflugerville     1104 Strickling Dr    78660   
2  2084491383  pflugerville    1408 Fort Dessau Rd    78660   
3   120901374  pflugerville     1025 Strickling Dr    78660   
4    60134862  pflugerville  15005 Donna Jane Loop    78660   

                                         description   latitude  longitude  \
0  14424 Lake Victor Dr, Pflugerville, TX 78660 i...  30.430632 -97.663078   
1  Absolutely GORGEOUS 4 Bedroom home with 2 full...  30.432672 -97.661697   
2  Under construction - estimated completion in A...  30.409748 -97.639771   
3  Absolutely darling one story home in charming ...  30.432112 -97.661659   
4  Brimming with appeal & warm livability! Sleek ...  30.437368 -97.656860   

   propertyTaxRate  garageSpaces  hasAssociation  hasCooling  hasGarage  \
0             1.98             2            True        True       True   
1             1.98             2            True        True       True   
2             1.98             0            True        True      False   
3             1.98             2            True        True       True   
4             1.98             0            True        True      False   

   hasHeating  hasSpa  hasView       homeType  parkingSpaces  yearBuilt  \
0        True   False    False  Single Family              2       2012   
1        True   False    False  Single Family              2       2013   
2        True   False    False  Single Family              0       2018   
3        True   False    False  Single Family              2       2013   
4        True   False    False  Single Family              0       2002   

   latestPrice  numPriceChanges latest_saledate  latest_salemonth  \
0       305000                5        9/2/2019                 9   
1       295000                1      10/13/2020                10   
2       256125                1       7/31/2019                 7   
3       240000                4        8/8/2018                 8   
4       239900                3      10/31/2018                10   

   latest_saleyear                                latestPriceSource  \
0             2019  Coldwell Banker United, Realtors - South Austin   
1             2020                                   Agent Provided   
2             2019                                   Agent Provided   
3             2018                                   Agent Provided   
4             2018                                   Agent Provided   

   numOfPhotos  numOfAccessibilityFeatures  numOfAppliances  \
0           39                           0                5   
1           29                           0                1   
2            2                           0                4   
3            9                           0                0   
4           27                           0                0   

   numOfParkingFeatures  numOfPatioAndPorchFeatures  numOfSecurityFeatures  \
0                     2                           1                      3   
1                     2                           0                      0   
2                     1                           0                      1   
3                     2                           0                      0   
4                     1                           0                      0   

   numOfWaterfrontFeatures  numOfWindowFeatures  numOfCommunityFeatures  \
0                        0                    1                       0   
1                        0                    0                       0   
2                        0                    0                       0   
3                        0                    0                       0   
4                        0                    0                       0   

   lotSizeSqFt  livingAreaSqFt  numOfPrimarySchools  numOfElementarySchools  \
0         6011            2601         

## b. Dropping Columns

In [13]:
# Dropping columns that are either irrelevant to analysis or contain PII data, will keep lat/long info for geographical component

df_house_2 = df_house.drop(['zpid', 'city', 'streetAddress', 'zipcode', 'description', 'homeImage'], axis = 1)

In [16]:
# Dropping parkingSpaces column - redundant since there is a garageSpaces column

df_house_3 = df_house_2.drop(columns = ['parkingSpaces'])

In [18]:
# Double checking columns

pd.options.display.max_columns = None
df_house_3.head()

latitude  longitude  propertyTaxRate  garageSpaces  hasAssociation  \
0  30.430632 -97.663078             1.98             2            True   
1  30.432672 -97.661697             1.98             2            True   
2  30.409748 -97.639771             1.98             0            True   
3  30.432112 -97.661659             1.98             2            True   
4  30.437368 -97.656860             1.98             0            True   

   hasCooling  hasGarage  hasHeating  hasSpa  hasView       homeType  \
0        True       True        True   False    False  Single Family   
1        True       True        True   False    False  Single Family   
2        True      False        True   False    False  Single Family   
3        True       True        True   False    False  Single Family   
4        True      False        True   False    False  Single Family   

   yearBuilt  latestPrice  numPriceChanges latest_saledate  latest_salemonth  \
0       2012       305000                5        9/2/2019                 9   
1       2013       295000                1      10/13/2020                10   
2       2018       256125                1       7/31/2019                 7   
3       2013       240000                4        8/8/2018                 8   
4       2002       239900                3      10/31/2018                10   

   latest_saleyear                                latestPriceSource  \
0             2019  Coldwell Banker United, Realtors - South Austin   
1             2020                                   Agent Provided   
2             2019                                   Agent Provided   
3             2018                                   Agent Provided   
4             2018                                   Agent Provided   

   numOfPhotos  numOfAccessibilityFeatures  numOfAppliances  \
0           39                           0                5   
1           29                           0                1   
2            2                           0                4   
3            9                           0                0   
4           27                           0                0   

   numOfParkingFeatures  numOfPatioAndPorchFeatures  numOfSecurityFeatures  \
0                     2                           1                      3   
1                     2                           0                      0   
2                     1                           0                      1   
3                     2                           0                      0   
4                     1                           0                      0   

   numOfWaterfrontFeatures  numOfWindowFeatures  numOfCommunityFeatures  \
0                        0                    1                       0   
1                        0                    0                       0   
2                        0                    0                       0   
3                        0                    0                       0   
4                        0                    0                       0   

   lotSizeSqFt  livingAreaSqFt  numOfPrimarySchools  numOfElementarySchools  \
0         6011            2601                    1                       0   
1         6185            1768                    1                       0   
2         7840            1478                    0                       2   
3         6098            1678                    1                       0   
4         6708            2132                    1                       0   

   numOfMiddleSchools  numOfHighSchools  avgSchoolDistance  avgSchoolRating  \
0                   1                 1           1.266667         2.666667   
1                   1                 1           1.400000         2.666667   
2                   1                 1           1.200000         3.000000   
3                   1                 1           1.400000         2.666667   
4                   1                 1          

In [19]:
# Double checking column numbers

df_house_3.shape

(15171, 40)

## c. Renaming Columns

In [26]:
# Changing latestPrice to price

df_house_3.rename(columns = {'latestPrice' : 'price'}, inplace = True)

In [27]:
# Changing latest_saledate to saleDate

df_house_3.rename(columns = {'latest_saledate' : 'saleDate'}, inplace = True)

In [28]:
# Changing latest_salemonth to saleMonth

df_house_3.rename(columns = {'latest_salemonth' : 'saleMonth'}, inplace = True)

In [29]:
# Changing sale_year to saleYear

df_house_3.rename(columns = {'latest_saleyear' : 'saleYear'}, inplace = True)

In [30]:
# Changing MedianStudentsPerTeacher to medianStudentsPerTeacher - keep header naming formating consistent

df_house_3.rename(columns = {'MedianStudentsPerTeacher' : 'medianStudentsPerTeacher'}, inplace = True)

In [31]:
# Double checking header names

df_house_3.head()

latitude  longitude  propertyTaxRate  garageSpaces  hasAssociation  \
0  30.430632 -97.663078             1.98             2            True   
1  30.432672 -97.661697             1.98             2            True   
2  30.409748 -97.639771             1.98             0            True   
3  30.432112 -97.661659             1.98             2            True   
4  30.437368 -97.656860             1.98             0            True   

   hasCooling  hasGarage  hasHeating  hasSpa  hasView       homeType  \
0        True       True        True   False    False  Single Family   
1        True       True        True   False    False  Single Family   
2        True      False        True   False    False  Single Family   
3        True       True        True   False    False  Single Family   
4        True      False        True   False    False  Single Family   

   yearBuilt   price  numPriceChanges    saleDate  saleMonth  saleYear  \
0       2012  305000                5    9/2/2019          9      2019   
1       2013  295000                1  10/13/2020         10      2020   
2       2018  256125                1   7/31/2019          7      2019   
3       2013  240000                4    8/8/2018          8      2018   
4       2002  239900                3  10/31/2018         10      2018   

                                 latestPriceSource  numOfPhotos  \
0  Coldwell Banker United, Realtors - South Austin           39   
1                                   Agent Provided           29   
2                                   Agent Provided            2   
3                                   Agent Provided            9   
4                                   Agent Provided           27   

   numOfAccessibilityFeatures  numOfAppliances  numOfParkingFeatures  \
0                           0                5                     2   
1                           0                1                     2   
2                           0                4                     1   
3                           0                0                     2   
4                           0                0                     1   

   numOfPatioAndPorchFeatures  numOfSecurityFeatures  numOfWaterfrontFeatures  \
0                           1                      3                        0   
1                           0                      0                        0   
2                           0                      1                        0   
3                           0                      0                        0   
4                           0                      0                        0   

   numOfWindowFeatures  numOfCommunityFeatures  lotSizeSqFt  livingAreaSqFt  \
0                    1                       0         6011            2601   
1                    0                       0         6185            1768   
2                    0                       0         7840            1478   
3                    0                       0         6098            1678   
4                    0                       0         6708            2132   

   numOfPrimarySchools  numOfElementarySchools  numOfMiddleSchools  \
0                    1                       0                   1   
1                    1                       0                   1   
2                    0                       2                   1   
3                    1                       0                   1   
4                    1                       0                   1   

   numOfHighSchools  avgSchoolDistance  avgSchoolRating  avgSchoolSize  \
0                 1           1.266667         2.666667           1063   
1                 1           1.400000         2.666667           1063   
2                 1           1.200000         3.000000           1108   
3                 1           1.400000         2.666667           1063   
4                 1           1.133333         4.000000           1223   

   medianStudents

# 04. Data Cleaning

## a. Duplicates

In [32]:
# Finding duplicate values

df_house_3[df_house_3.duplicated()]

Empty DataFrame
Columns: [latitude, longitude, propertyTaxRate, garageSpaces, hasAssociation, hasCooling, hasGarage, hasHeating, hasSpa, hasView, homeType, yearBuilt, price, numPriceChanges, saleDate, saleMonth, saleYear, latestPriceSource, numOfPhotos, numOfAccessibilityFeatures, numOfAppliances, numOfParkingFeatures, numOfPatioAndPorchFeatures, numOfSecurityFeatures, numOfWaterfrontFeatures, numOfWindowFeatures, numOfCommunityFeatures, lotSizeSqFt, livingAreaSqFt, numOfPrimarySchools, numOfElementarySchools, numOfMiddleSchools, numOfHighSchools, avgSchoolDistance, avgSchoolRating, avgSchoolSize, medianStudentsPerTeacher, numOfBathrooms, numOfBedrooms, numOfStories]
Index: []

###### No duplicates found

## b. Mixed Data Types

In [33]:
# Checking mixed data types

for col in df_house_3.columns.tolist():
  weird = (df_house_3[[col]].applymap(type) != df_house_3[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_house_3[weird]) > 0:
    print (col)

###### No columns printed out

## c. Missing Data

In [34]:
# Checking for missing values

df_house_3.isnull().sum()

latitude                      0
longitude                     0
propertyTaxRate               0
garageSpaces                  0
hasAssociation                0
hasCooling                    0
hasGarage                     0
hasHeating                    0
hasSpa                        0
hasView                       0
homeType                      0
yearBuilt                     0
price                         0
numPriceChanges               0
saleDate                      0
saleMonth                     0
saleYear                      0
latestPriceSource             0
numOfPhotos                   0
numOfAccessibilityFeatures    0
numOfAppliances               0
numOfParkingFeatures          0
numOfPatioAndPorchFeatures    0
numOfSecurityFeatures         0
numOfWaterfrontFeatures       0
numOfWindowFeatures           0
numOfCommunityFeatures        0
lotSizeSqFt                   0
livingAreaSqFt                0
numOfPrimarySchools           0
numOfElementarySchools        0
numOfMid

###### No missing values in data set

# 05. Basic Descriptive Stats

In [35]:
# Conducting descriptive analysis

df_house_3.describe()

latitude     longitude  propertyTaxRate  garageSpaces  \
count  15171.000000  15171.000000     15171.000000  15171.000000   
mean      30.291596    -97.778532         1.994085      1.229187   
std        0.096973      0.084715         0.053102      1.352117   
min       30.085030    -98.022057         1.980000      0.000000   
25%       30.203313    -97.838009         1.980000      0.000000   
50%       30.284416    -97.769539         1.980000      1.000000   
75%       30.366585    -97.717903         1.980000      2.000000   
max       30.517323    -97.569504         2.210000     22.000000   

          yearBuilt         price  numPriceChanges     saleMonth  \
count  15171.000000  1.517100e+04     15171.000000  15171.000000   
mean    1988.554743  5.127677e+05         3.032892      6.733043   
std       21.579458  4.531853e+05         2.487392      3.146760   
min     1905.000000  5.500000e+03         1.000000      1.000000   
25%     1974.000000  3.090000e+05         1.000000      4.000000   
50%     1993.000000  4.050000e+05         2.000000      7.000000   
75%     2006.000000  5.750000e+05         4.000000      9.000000   
max     2020.000000  1.350000e+07        23.000000     12.000000   

           saleYear   numOfPhotos  numOfAccessibilityFeatures  \
count  15171.000000  15171.000000                15171.000000   
mean    2019.078241     30.640432                    0.012985   
std        0.813941     21.175174                    0.179438   
min     2018.000000      1.000000                    0.000000   
25%     2018.000000     20.000000                    0.000000   
50%     2019.000000     29.000000                    0.000000   
75%     2020.000000     39.000000                    0.000000   
max     2021.000000    179.000000                    8.000000   

       numOfAppliances  numOfParkingFeatures  numOfPatioAndPorchFeatures  \
count     15171.000000          15171.000000                15171.000000   
mean          3.474722              1.710105                    0.663371   
std           1.893351              0.788641                    0.983779   
min           0.000000              0.000000                    0.000000   
25%           2.000000              1.000000                    0.000000   
50%           3.000000              2.000000                    0.000000   
75%           4.000000              2.000000                    1.000000   
max          12.000000              6.000000                    8.000000   

       numOfSecurityFeatures  numOfWaterfrontFeatures  numOfWindowFeatures  \
count           15171.000000             15171.000000         15171.000000   
mean                0.466878                 0.002768             0.208490   
std                 0.824481                 0.061771             0.488331   
min                 0.000000                 0.000000             0.000000   
25%                 0.000000                 0.000000             0.000000   
50%                 0.000000                 0.000000             0.000000   
75%                 1.000000                 0.000000             0.000000   
max                 6.000000                 2.000000             4.000000   

       numOfCommunityFeatures   lotSizeSqFt  livingAreaSqFt  \
count            15171.000000  1.517100e+04    15171.000000   
mean                 0.018852  1.190839e+05     2208.323314   
std                  0.206775  1.225272e+07     1378.576119   
min                  0.000000  1.000000e+02      300.000000   
25%                  0.000000  6.534000e+03     1483.000000   
50%                  0.000000  8.276000e+03     1975.000000   
75%                  0.000000  1.089000e+04     2687.000000   
max                  8.000000  1.508483e+09   109292.000000   

       numOfPrimarySchools  numOfElementarySchools  numOfMiddleSchools  \
count         15171.000000            15171.000000        15171.000000   
mean              0.940676                0.049173            1.035990   
std               0.2

In [36]:
# Viewing in Excel

df_house_3.describe().to_clipboard()

###### Some of the numbers listed are raising some questions. The first one being the max amount of garage space being 22?

In [38]:
# Checking all values for garageSpaces column

df_house_3['garageSpaces'].value_counts(dropna = False)

0     6826
2     5508
1     1098
3      938
4      581
6       99
5       74
8       17
7       11
10       9
12       5
9        3
20       1
22       1
Name: garageSpaces, dtype: int64

###### I checked the original data source, zillow.com, and it appears the web scraper took the values in parkingSpaces for garageSpaces as well since the garageSpaces variable is not explicitly stated. parkingSpaces is a broad term and can entail different definitions which are not clearly explained. The column will be dropped due to the possibility of inaccurate data.

In [39]:
# Dropping garageSpaces column

df_house_4 = df_house_3.drop(columns = ['garageSpaces'])

In [40]:
df_house_4.head(1)

latitude  longitude  propertyTaxRate  hasAssociation  hasCooling  \
0  30.430632 -97.663078             1.98            True        True   

   hasGarage  hasHeating  hasSpa  hasView       homeType  yearBuilt   price  \
0       True        True   False    False  Single Family       2012  305000   

   numPriceChanges  saleDate  saleMonth  saleYear  \
0                5  9/2/2019          9      2019   

                                 latestPriceSource  numOfPhotos  \
0  Coldwell Banker United, Realtors - South Austin           39   

   numOfAccessibilityFeatures  numOfAppliances  numOfParkingFeatures  \
0                           0                5                     2   

   numOfPatioAndPorchFeatures  numOfSecurityFeatures  numOfWaterfrontFeatures  \
0                           1                      3                        0   

   numOfWindowFeatures  numOfCommunityFeatures  lotSizeSqFt  livingAreaSqFt  \
0                    1                       0         6011            2601   

   numOfPrimarySchools  numOfElementarySchools  numOfMiddleSchools  \
0                    1                       0                   1   

   numOfHighSchools  avgSchoolDistance  avgSchoolRating  avgSchoolSize  \
0                 1           1.266667         2.666667           1063   

   medianStudentsPerTeacher  numOfBathrooms  numOfBedrooms  numOfStories  
0                        14             3.0              4             2

###### numOfBathrooms maximum is 27, this is a bit odd

In [41]:
# Checking numOfBathrooms values

df_house_4['numOfBathrooms'].value_counts(dropna = False)

2.00     6101
3.00     5280
4.00     1884
1.00      932
5.00      408
6.00      157
2.50      126
0.00      125
7.00       61
3.50       44
8.00       17
1.50       13
10.00       5
4.50        5
1.75        3
5.50        2
2.75        2
9.00        2
6.50        1
13.00       1
27.00       1
1.70        1
Name: numOfBathrooms, dtype: int64

###### Bathroom values can end in .0, .5, or .75. I am assuming 1.70 was a mistake and will replace it with 2 since it is a more common value. 

In [46]:
# Replacing 1.70 value with 2.00

df_house_4.loc[df_house_4['numOfBathrooms'] == 1.70, 'numOfBathrooms'] = 2.00

In [47]:
# Checking numOfBathrooms values

df_house_4['numOfBathrooms'].value_counts(dropna = False)

2.00     6102
3.00     5280
4.00     1884
1.00      932
5.00      408
6.00      157
2.50      126
0.00      125
7.00       61
3.50       44
8.00       17
1.50       13
10.00       5
4.50        5
1.75        3
9.00        2
2.75        2
5.50        2
13.00       1
27.00       1
6.50        1
Name: numOfBathrooms, dtype: int64

###### 27 may have been a typo for either: 2.75, 2.50, or 3.00. I will go with 3 since it is more common compared to the other two values

In [48]:
# Replacing 27 value with 3

df_house_4.loc[df_house_4['numOfBathrooms'] == 27.00, 'numOfBathrooms'] = 3.00

###### There are 125 '0' values, this needs to be explored further

In [49]:
# Looking for rows with 0 bathrooms listed

df_house_4.loc[df_house_4['numOfBathrooms'] == 0]

latitude  longitude  propertyTaxRate  hasAssociation  hasCooling  \
311    30.332357 -97.858322             1.98            True        True   
346    30.364504 -97.787285             1.98           False        True   
705    30.323517 -97.861473             1.98            True        True   
964    30.329847 -97.926910             1.98           False       False   
1023   30.342888 -97.863258             1.98            True        True   
...          ...        ...              ...             ...         ...   
14512  30.330925 -97.765823             1.98           False        True   
14622  30.295774 -97.745300             1.98           False        True   
14682  30.314917 -97.743362             1.98           False       False   
14716  30.294628 -97.749329             1.98           False        True   
14789  30.317568 -97.745750             1.98           False        True   

       hasGarage  hasHeating  hasSpa  hasView       homeType  yearBuilt  \
311         True        True   False     True    Vacant Land       2020   
346        False        True   False    False    Vacant Land       1960   
705        False        True   False    False  Single Family       2012   
964        False       False   False    False    Vacant Land       2018   
1023       False        True   False    False    Vacant Land       1965   
...          ...         ...     ...      ...            ...        ...   
14512       True        True   False     True    Vacant Land       1948   
14622       True        True   False    False    Vacant Land       1924   
14682      False        True   False    False    Vacant Land       1936   
14716      False        True   False     True  Single Family       1938   
14789       True        True   False     True    Vacant Land       1954   

         price  numPriceChanges    saleDate  saleMonth  saleYear  \
311     625000               11   9/28/2020          9      2020   
346     685000                3   6/30/2019          6      2019   
705     385000                2  11/12/2018         11      2018   
964     139990                1    4/7/2018          4      2018   
1023    322500                2   1/24/2020          1      2020   
...        ...              ...         ...        ...       ...   
14512  1250000                5  12/14/2020         12      2020   
14622   465000                4  12/21/2018         12      2018   
14682   450000                1  12/14/2018         12      2018   
14716  1585000                4    1/9/2020          1      2020   
14789   600000                3   3/13/2019          3      2019   

                                    latestPriceSource  numOfPhotos  \
311                                   Broker Provided           15   
346     Coldwell Banker United, Realtors - 183 Office           29   
705                                    Agent Provided            1   
964                                    Agent Provided            3   
1023                                   Agent Provided           31   
...                                               ...          ...   
14512                                  Agent Provided           50   
14622                                 Broker Provided           13   
14682                                  Agent Provided           25   
14716                                  Agent Provided           35   
14789  Keller Williams Realty Southwest Market Center            2   

       numOfAccessibilityFeatures  numOfAppliances  numOfParkingFeatures  \
311                             0                4                     2   
346                             0                0                     1   
705                             0                0                     1   
964                             0                0                     1   
1023                            0                4                     1   
...                           ...              ...                   ...   
1451

###### It appears a lot of the 0 bathrooms belong to Vacant Land property type

In [50]:
# Checking homeType values

df_house_4['homeType'].value_counts(dropna = False)

Single Family            14241
Condo                      470
Townhouse                  174
Multiple Occupancy          96
Vacant Land                 83
Residential                 37
Apartment                   37
Mobile / Manufactured       17
MultiFamily                 10
Other                        6
Name: homeType, dtype: int64

In [51]:
# Creating new df without 'Vacant Land' & 'Other' - only interested in land being used; unsure what 'Other' would consist of

df_house_5 = df_house_4[(df_house_4['homeType']!='Vacant Land') & (df_house_4['homeType']!='Other')]

In [52]:
# Double checking homeType values

df_house_5['homeType'].value_counts(dropna = False)

Single Family            14241
Condo                      470
Townhouse                  174
Multiple Occupancy          96
Residential                 37
Apartment                   37
Mobile / Manufactured       17
MultiFamily                 10
Name: homeType, dtype: int64

In [53]:
# Re-checking rows with 0 bathrooms listed

df_house_5.loc[df_house_5['numOfBathrooms'] == 0]

latitude  longitude  propertyTaxRate  hasAssociation  hasCooling  \
705    30.323517 -97.861473             1.98            True        True   
1842   30.312716 -97.900482             1.98            True        True   
2223   30.390335 -97.882072             1.98            True        True   
2557   30.345898 -97.859985             1.98            True        True   
3328   30.375504 -97.688255             1.98           False        True   
3470   30.357208 -97.679146             1.98           False       False   
3496   30.372677 -97.718483             1.98            True        True   
3852   30.426914 -97.759666             1.98           False        True   
4008   30.234682 -97.692123             1.98           False        True   
4108   30.223991 -97.667549             1.98           False        True   
4464   30.199942 -97.984703             2.01            True        True   
4491   30.200550 -97.969345             2.01            True        True   
5150   30.191986 -97.812904             1.98            True        True   
5404   30.253592 -97.894714             1.98           False        True   
5466   30.211281 -97.780289             1.98           False        True   
6252   30.269354 -97.818771             1.98            True        True   
6615   30.182751 -97.834686             1.98           False        True   
6646   30.188589 -97.823662             1.98           False        True   
6944   30.200434 -97.809883             1.98           False        True   
6956   30.197086 -97.805405             1.98           False        True   
7034   30.205379 -97.804886             1.98           False        True   
7040   30.199633 -97.786552             1.98           False        True   
7049   30.204607 -97.808357             1.98           False       False   
8021   30.376259 -97.750069             1.98           False        True   
8232   30.497154 -97.760078             2.21            True        True   
8633   30.413637 -97.730003             1.98           False        True   
10009  30.291788 -97.682999             1.98           False        True   
10083  30.256399 -97.723633             1.98           False        True   
10257  30.262421 -97.687126             1.98           False       False   
10362  30.192862 -97.730164             1.98           False        True   
10382  30.201094 -97.748589             1.98           False        True   
10779  30.228161 -97.742615             1.98           False        True   
10790  30.228567 -97.737854             1.98           False        True   
11176  30.312010 -97.710106             1.98           False       False   
11215  30.295963 -97.716385             1.98           False        True   
11625  30.305782 -97.679497             1.98           False        True   
11894  30.337627 -97.712486             1.98           False        True   
11997  30.178684 -97.834274             1.98           False        True   
12003  30.164707 -97.834358             1.98            True        True   
12316  30.210625 -97.871208             1.98            True        True   
12993  30.197540 -97.862419             1.98            True        True   
12995  30.274494 -97.760086             1.98           False        True   
13831  30.238840 -97.781548             1.98           False        True   
13848  30.245035 -97.761261             1.98           False        True   
14056  30.460995 -97.759834             2.21            True        True   
14313  30.299656 -97.781372             1.98           False        True   
14357  30.298880 -97.767937             1.98           False        True   
14716  30.294628 -97.749329             1.98           False        True   

       hasGarage  hasHeating  hasSpa  hasView            homeType  yearBuilt  \
705        False        True   False    False       Single Family       2012   
1842        True        True   False     True       Single Family       2011   
2223        True        True   False     Tr

In [54]:
# Viewing in Excel

df_house_5.loc[df_house_5['numOfBathrooms'] == 0].to_clipboard()

###### Only 48 rows contain 0 bathrooms. I am assuming the data was unavailable or missing on Zillow. These rows will be removed since they make up less than 1% of total data

In [55]:
# Creating new df w/ bathroom values

df_house_6 = df_house_5.loc[df_house_5['numOfBathrooms'] > 0]

In [56]:
df_house_6.shape

(15034, 39)

In [57]:
# Checking numOfBedroom values - original descriptive analysis showed minimum was 0

df_house_6['numOfBedrooms'].value_counts(dropna = False)

3     7402
4     5169
5     1161
2     1048
6      120
1       91
8       18
7       13
0       10
10       1
20       1
Name: numOfBedrooms, dtype: int64

In [58]:
# Taking a closer look into 0 numOfBedroom values

df_house_6.loc[df_house_5['numOfBedrooms'] == 0]

latitude  longitude  propertyTaxRate  hasAssociation  hasCooling  \
5098   30.189238 -97.831108             1.98           False        True   
8607   30.432478 -97.712067             1.98           False        True   
8759   30.441351 -97.684914             1.98            True        True   
9685   30.272972 -97.716553             1.98           False        True   
10006  30.281124 -97.692078             1.98           False        True   
10698  30.231682 -97.758957             1.98            True        True   
10780  30.225664 -97.754723             1.98           False        True   
13314  30.284014 -97.762421             1.98            True        True   
13601  30.197281 -97.901817             1.98            True        True   
14972  30.343782 -97.732636             1.98           False        True   

       hasGarage  hasHeating  hasSpa  hasView       homeType  yearBuilt  \
5098        True        True   False     True  Single Family       1983   
8607        True        True   False    False  Single Family       1982   
8759       False        True   False     True  Single Family       1984   
9685       False        True   False    False  Single Family       1929   
10006      False        True   False    False  Single Family       1930   
10698      False        True   False     True          Condo       2016   
10780      False        True   False    False      Apartment       1973   
13314      False        True   False    False          Condo       1972   
13601       True        True   False    False  Single Family       2007   
14972      False        True   False    False  Single Family       1951   

        price  numPriceChanges    saleDate  saleMonth  saleYear  \
5098   298900                2    4/6/2020          4      2020   
8607   525000                1   5/22/2019          5      2019   
8759   234000                1   7/12/2019          7      2019   
9685   349000                5  12/16/2019         12      2019   
10006  430000                3   12/2/2019         12      2019   
10698  224900                1    5/6/2019          5      2019   
10780  605000                8   5/16/2018          5      2018   
13314  175000                1   10/1/2019         10      2019   
13601  469000                1   4/30/2019          4      2019   
14972  420000                1   10/3/2019         10      2019   

             latestPriceSource  numOfPhotos  numOfAccessibilityFeatures  \
5098           Broker Provided           28                           0   
8607            Agent Provided           24                           0   
8759           Broker Provided           34                           0   
9685   Luisa Mauro Real Estate           20                           0   
10006           Agent Provided           16                           0   
10698           Agent Provided           17                           0   
10780           Agent Provided           32                           0   
13314           Agent Provided           16                           0   
13601          Broker Provided            1                           0   
14972           Agent Provided            9                           0   

       numOfAppliances  numOfParkingFeatures  numOfPatioAndPorchFeatures  \
5098                 4                     2                           3   
8607                 5                     3                           0   
8759                 3                     1                           2   
9685                 2                     1                           0   
10006                4                     1                           0   
10698                7                     0                           0   
10780                0                     2                           0   
13314                3                     0                           1   
13601                4                     2                           0   
14972                

###### Only 10 rows contain 0 bedrooms, will remove these as well

In [61]:
# Creating new df w/ bedroom values

df_house_7 = df_house_6.loc[df_house_6['numOfBedrooms'] > 0]

In [62]:
df_house_7.shape

(15024, 39)

In [64]:
# Checking bedroom maximum value of 20

df_house_7.loc[df_house_7['numOfBedrooms'] == 20]

latitude  longitude  propertyTaxRate  hasAssociation  hasCooling  \
8597  30.415489 -97.737213             1.98            True        True   

      hasGarage  hasHeating  hasSpa  hasView homeType  yearBuilt   price  \
8597      False        True   False     True    Condo       1985  174900   

      numPriceChanges    saleDate  saleMonth  saleYear latestPriceSource  \
8597                2  11/13/2020         11      2020    Agent Provided   

      numOfPhotos  numOfAccessibilityFeatures  numOfAppliances  \
8597           25                           0                4   

      numOfParkingFeatures  numOfPatioAndPorchFeatures  numOfSecurityFeatures  \
8597                     1                           0                      0   

      numOfWaterfrontFeatures  numOfWindowFeatures  numOfCommunityFeatures  \
8597                        0                    0                       0   

      lotSizeSqFt  livingAreaSqFt  numOfPrimarySchools  \
8597         5227             916                    1   

      numOfElementarySchools  numOfMiddleSchools  numOfHighSchools  \
8597                       0                   1                 1   

      avgSchoolDistance  avgSchoolRating  avgSchoolSize  \
8597                2.4              6.0           1474   

      medianStudentsPerTeacher  numOfBathrooms  numOfBedrooms  numOfStories  
8597                        15             2.0             20             1

###### The squarefoot seems too small for there to be 20 bedrooms. It also does not make sense there would only be 2 bathrooms for 20 bedrooms. I will assume this is a typo and will replace 20 with 2.

In [66]:
# Replacing bedroom value of 20 with 2

df_house_7.loc[df_house_7['numOfBedrooms'] == 20, 'numOfBedrooms'] = 2

# 06. Exporting Data

In [74]:
# Exporting as csv file

df_house_7.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'austin_housing_clean.csv'))